async_hooks: C++ Embedder API overhaul

AndreasMadsen opened the pull request:
##### Checklist

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
async_hooks

I'm not sure how to split up the `AsyncHooksGetTriggerAsyncId` and the `async_context` commit. If someone has a specific suggestion I would be happy to do it.

* Fix `AsyncHooksGetTriggerAsyncId` such it corresponds to `async_hooks.triggerAsyncId` and not `async_hooks.initTriggerId`.
* Use an `async_context` struct instead of two `async_uid` values. This change was necessary since the fixing `AsyncHooksGetTriggerAsyncId` otherwise makes it impossible to get the correct default trigger id. It also prevents an invalid `triggerAsyncId` in `MakeCallback`.
* Rename `async_uid` to `async_id` for consistency.

/cc @nodejs/async_hooks

AndreasMadsen commented:
I think this is a semver-minor since it only touches `async_hooks` experimental APIs.

addaleax commented on src/node.h:
```diff
@@ -517,7 +517,11 @@ typedef void (*promise_hook_func) (v8::PromiseHookType type,
 
-typedef double async_uid;
+typedef double async_id;
+typedef struct async_context {
```
Fwiw you don’t need a typedef, just `struct async_context { ... };` will work fine.

refack commented:
Does this solve https://github.com/nodejs/help/issues/644?

addaleax commented:
@refack It’s unlikely, and it’s not really clear what the problem in that issue is in the first place.

refack commented:
> @refack It’s unlikely, and it’s not really clear what the problem in that issue is in the first place.

AFAICT the issue there is `Environment::AsyncHooks::ExecScope` being internal, but necessary in order to start the loop.

addaleax commented:
@refack Can we keep the discussion for that issue in that issue? What you’re saying sounds reasonable, could maybe post a minimal example?

refack commented:
> @refack Can we keep the discussion for that issue in that issue? What you’re saying sounds reasonable, could maybe post a minimal example?

Sure.

addaleax commented on src/node.h:
```diff
@@ -548,13 +552,14 @@ NODE_EXTERN NODE_DEPRECATED("Use AsyncHooksGetTriggerAsyncId(isolate)",
+NODE_EXTERN async_context EmitAsyncInit(v8::Isolate* isolate,
+                                        v8::Local<v8::Object> resource,
+                                        const char* name,
+                                        async_id trigger_async_id = -1);
```
It seems to me like this breakage would be shim-able… If that’s right, I’d say we label this dont-land and I can do a non-breaking backport to v8.x, if you like

AndreasMadsen replied:
That would be awesome. In particular, I couldn't see how to make `EmitAsyncInit` backward compatible.

addaleax replied:
The old API required `trigger_async_id` instead of leaving it optional, so we could make it return a pair based on that parameter, right?

AndreasMadsen replied:
What should `EmitAsyncInit` then return when `trigger_async_id` is specified?

addaleax replied:
I don’t really understand the question, sorry. Basically, what we need to shim is the old definition
```c++
async_uid EmitAsyncInit(v8::Isolate* isolate,
                        v8::Local<v8::Object> resource,
                        const char* name,
                        async_uid trigger_async_id);
```
in terms of the new one in this PR, right?

So what’s speaking against `OldStyleEmitAsyncInit(resource, name, trigger_async_id) := (NewStyleAsyncInit(resource, name, trigger_async_id)).async_id_`?

Or are you thinking about the return type? The hacky solution would be to enable casting from `async_context` to `async_id` by returning `async_id_` (and immediately deprecating that), it’s not pretty but as long as it doesn’t end up in master I’m okay with that.

(One more thing: I’m noticing, maybe we should not underscore-suffix the fields of `async_context` given that they are public, not internal…)

AndreasMadsen replied:
> Or are you thinking about the return type? The hacky solution would be to enable casting from `async_context` to `async_id` by returning `async_id_` (and immediately deprecating that), it’s not pretty but as long as it doesn’t end up in master I’m okay with that.

Yes, it is the return type. The casting makes sense.

> (One more thing: I’m noticing, maybe we should not underscore-suffix the fields of `async_context` given that they are public, not internal…)

Right. Please check ffdd431d7d42efe939f3cc378814844b5796e4b5.

AndreasMadsen commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/8944/

AndreasMadsen commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/8968/

trevnorris commented on src/node.h:
```diff
@@ -637,21 +636,20 @@ class AsyncResource {
 
-    async_uid get_uid() const {
-      return uid_;
+    async_id get_uid() const {
```
For the sake of keeping things the same, want to rename this to `get_async_id()`?

AndreasMadsen commented:
Rebased and added `get_trigger_async_id` to match the JavaScript version of `AsyncResource`.

CI: https://ci.nodejs.org/job/node-test-pull-request/8993/

AndreasMadsen commented:
landed in c6ce500edf364692efa9d46bc1bd9e959611f7da
